File: counting_sort.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

str = "abracadabra"
freq = [0] * 0xff
cum  = [0] * 0xff
work = [0] * len(str)

for x in map(lambda x: ord(x), str):
    freq[x] += 1

for i in xrange(1, 0xff):
    cum[i] = cum[i - 1] + freq[i]

for i in xrange(len(str) - 1, -1, -1):
    cum[ord(str[i])] -= 1
    work[cum[ord(str[i])]] = str[i]

print work

File: counting_sort.pl
#!/usr/bin/env perl
use strict;
use warnings;

use FindBin::libs;
use Perl6::Say;

use constant UCHAR_MAX => 0x100;

my $str = "abracadabra\$";
my @chars = unpack('C*', $str);

my @freq;
for (my $i = 0; $i < UCHAR_MAX; $i++) {
    $freq[$i] = 0;
}

for (@chars) {
    $freq[$_]++;
}

say @freq;

my @cum = (0);
for (my $i = 0; $i < UCHAR_MAX; $i++) {
    $cum[$i] = $cum[$i - 1] + $freq[$i];
}

say @cum;

my @sorted;
my $n = @chars;
for (my $i = $n - 1; $i >= 0; $i--) {
    $sorted[ --$cum[ $chars[$i] ] ] = $i;
}

# say chr for @sorted;
say for @sorted;
